modal/_utils/async_utils.py

```diff
@@ -463,3 +476,93 @@ def run_generator_sync(
         except BaseException as err:
             exc = err
     loop.close()
+
+
+@asynccontextmanager
+async def aclosing(gen: AsyncGenerator[T, None]) -> AsyncGenerator[AsyncGenerator[T, None], None]:
+    try:
+        yield gen
+    finally:
+        await gen.aclose()
+
+
+async def async_map(
+    input: AsyncIterator[T], async_mapper_func: Callable[[T], Awaitable[V]], concurrency: int, in_order: bool = False
+) -> AsyncIterator[V]:
+    input_queue: asyncio.Queue[T] = asyncio.Queue(maxsize=concurrency)
+    results_queue: asyncio.Queue[V] = asyncio.Queue()
+
+    # TODO: figure out how to return in order
+
+    async def producer():
+        async for item in input:
+            await input_queue.put(item)
+
+    async def worker():
+        while True:
+            item = await input_queue.get()
+            result = await async_mapper_func(item)
+            await results_queue.put(result)
+            input_queue.task_done()
+
+    producer_task = asyncio.create_task(producer())
+    worker_tasks = [asyncio.create_task(worker()) for _ in range(concurrency)]
+
+    async def complete_map():
+        await producer_task
+        await input_queue.join()
+
+    complete_map_task = asyncio.create_task(complete_map())
+
+    try:
+        while True:
+            await asyncio.wait([complete_map_task, producer_task, *worker_tasks], return_when=asyncio.FIRST_COMPLETED)
```

Comment:
This will block until one of these tasks has completed, which could be long after the first result is produced. You'd have to have a task that also checks if any new outputs have been created on the queue.
Either an `asyncio.create_task(queue.get()` or (maybe preferable) some separate `asyncio.Event` that is set whenever a worker puts an item and reset whenever you drain the output queue

Reply:
Make sure to add a test for this too. Should be testable by having a map workload that takes X time to complete and using 1 concurrent tasks you should be able to get 1 output roughly every X time